import sys
import os
import datetime
import re
from PySide6.QtWidgets import QApplication, QMessageBox
from PySide6.QtUiTools import QUiLoader
from PySide6.QtCore import QFile, QIODevice

from menu_logic import MenuManager
from translations import TEXTS  # <--- Importing texts

def resource_path(relative_path):
    # Obtains the absolute path to resources (required for PyInstaller)
    try:
        # PyInstaller creates a temporary folder named _MEIxxxx
        base_path = sys._MEIPASS
    except Exception:
        base_path = os.path.abspath(".")
    return os.path.join(base_path, relative_path)

def main():
    app = QApplication(sys.argv)

    # 1. INTERFACE DOWNLOAD
    ui_file_path = resource_path("ui.ui") 
    ui_file = QFile(ui_file_path) # Используем исправленную переменную
    
    if not ui_file.open(QIODevice.ReadOnly):
        # Печатаем ui_file_path, чтобы в терминале видеть, где он искал файл
        print(f"Ошибка: Не могу найти файл по пути {ui_file_path}")
        sys.exit(-1)
        
    loader = QUiLoader()
    window = loader.load(ui_file)
    ui_file.close()

    # --- MENU ---
    menu = MenuManager(window)
    # ------------------------

    # 2. LOGIC OF PRESERVATION
    def save_to_obsidian():
        # Get the current language from the window (it is changed by MenuManager)
        lang = getattr(window, 'current_lang', 'ru') 
        t = TEXTS[lang] # Current language dictionary

        try:
            text_pp = window.input_pp.toPlainText()
            text_pm = window.input_pm.toPlainText()
            text_mp = window.input_mp.toPlainText()
            text_mm = window.input_mm.toPlainText()
        except AttributeError:
            return

        def calculate_points(text):
            score = 0
            for line in text.split('\n'):
                line = line.strip()
                match = re.search(r'(?:^|\s)([1-5])$', line)
                if match: score += int(match.group(1))
            return score

        score_pp = calculate_points(text_pp)
        score_mm = calculate_points(text_mm)
        score_pm = calculate_points(text_pm)
        score_mp = calculate_points(text_mp)

        total_action = score_pp + score_mm
        total_status_quo = score_pm + score_mp

        # WE USE TRANSLATIONS FOR THE VERDICT
        if total_action > total_status_quo:
            verdict_text = t["verdict_do"]
        elif total_status_quo > total_action:
            verdict_text = t["verdict_stay"]
        else:
            verdict_text = t["verdict_draw"]

        # Putting together the verdict line
        final_verdict = f"{verdict_text} ({total_action} vs {total_status_quo})"

        timestamp = datetime.datetime.now().strftime("%d.%m.%Y %H:%M")
        
        # Generate a file using header translations
        content = f"""# {t['file_header']} {timestamp}
**ITOG:** {final_verdict}

---

## 1. {t['box_1']}
**{t['score']}: {score_pp}**

{text_pp if text_pp else "—"}

## 2. {t['box_2']}
**{t['score']}: {score_pm}**

{text_pm if text_pm else "—"}

## 3. {t['box_3']}
**{t['score']}: {score_mp}**

{text_mp if text_mp else "—"}

## 4. {t['box_4']}
**{t['score']}: {score_mm}**

{text_mm if text_mm else "—"}

---
*Generated by Descartes Decision Helper*
"""
        filename = f"decision_{datetime.datetime.now().strftime('%d.%m.%Y_%H-%M')}.txt"
        with open(filename, "w", encoding="utf-8") as f:
            f.write(content)
        
        # Success message in the desired language
        QMessageBox.information(window, t["saved_title"], f"{final_verdict}\n\n{t['saved_body']}:\n{filename}")

    try:
        window.btn_save.clicked.connect(save_to_obsidian)
    except AttributeError:
        pass

    window.show()
    sys.exit(app.exec())

if __name__ == "__main__":
    main()